src/content.ts

```diff
@@ -1,69 +1,121 @@
+export const broadcast = async (
+  contentOptions: activityPub.ActivityPubOpts,
+  opts?: config.ConfigOpts
+): Promise<batchMessages.BatchBroadcastMessage> => {
+  // Create the activity pub file
+  const contentObj = activityPub.create(contentOptions);
+  if (!activityPub.validate(contentObj)) throw InvalidActivityPubOpts;
+  const content = activityPub.serialize(contentObj);
+
+  // Upload the content file
+  const filename = getRandomString();
+  const uri = await storage.put(filename, content, opts);
+
+  // Get current user id
+  const { currentUserId } = await config.getConfig(opts);
+  if (!currentUserId) throw MissingUser;
+
+  // Creates and returns the DSNP Broadcast message
+  const contentHash = activityPub.hash(contentObj);
+  const message = messages.createBroadcastMessage(currentUserId, uri.toString(), contentHash);
+
+  // Sign and return the message
+  const signedMessage = await messages.sign(message, opts);
+  return signedMessage as batchMessages.BatchBroadcastMessage;
 };
 
 /**
- * reply() creates a reply activity pub event and enqueues it for the next
- * batch. This method is not yet implemented.
+ * reply() creates an activity pub file with the given content options,
+ * uploads it with a random filename using the configured storage adapter and
+ * creates a DSNP reply message for the hosted file for later announcement.
  *
- * @param content The content for the reply to broadcast
- * @param opts    Optional. Configuration overrides, such as from address, if any
- * @return The Batch Broadcast Message
+ * @param contentOptions Options for the activity pub content to generate
+ * @param inReplyTo      The DSNP Id of the message that this message is in reply to
+ * @param opts           Optional. Configuration overrides, such as from address, if any
+ * @returns              A Signed DSNP Reply message ready for inclusion in a batch
  */
-export const reply = async (_content: ActivityPubOpts, _opts?: ConfigOpts): Promise<BatchReplyMessage> => {
-  throw NotImplementedError;
-
-  // const config = config.getConfig(opts);
-  //
-  // const inReplyTo = activityPubOpts.inReplyTo;
-  // const activityPubObject = activityPub.create(activityPubOpts);
-  // const activityPubHash = activityPub.hash(activityPubObject);
-  // const uri = storage.store(activityPubObject, opts);
-  //
-  // return events.createReplyEvent(uri, inReplyTo, activityPubHash);
+export const reply = async (
+  contentOptions: activityPub.ActivityPubOpts,
+  inReplyTo: string,
+  opts?: config.ConfigOpts
+): Promise<batchMessages.BatchReplyMessage> => {
+  // Validate inReplyTo
+  if (!validateDSNPId(inReplyTo)) throw InvalidInReplyTo;
+
+  // Create the activity pub file
+  const contentObj = activityPub.create(contentOptions);
+  if (!activityPub.validateReply(contentObj)) throw InvalidActivityPubOpts;
+  const content = activityPub.serialize(contentObj);
+
+  // Upload the content file
+  const filename = getRandomString();
```

Comment:
I was thinking if a user really wants to use a specific filename they could call `storage.put()` directly instead of using the porcelain, but I'm not opposed to making it an option. I'm a little unsure how we would implement it though. If we made it an optional param like this...

```
export const reply = async (
  contentOptions: activityPub.ActivityPubOpts,
  inReplyTo: string,
  filename?: string,
  opts?: config.ConfigOpts
): Promise<batchMessages.BatchReplyMessage> => {
```

Then it becomes unclear whether a user meant to pass a filename or config opts when they do something like this...

```
const msg = await content.reply({...}, "dsnp://...", x);
```

Maybe instead there should be some kind of `filenameFunc` option in the config? Then the default could be `getRandomString` and users could override it if they prefer a different convention?

Reply:
I thought we were removing the "id" field requirement of ActivityPub so we could default this to the hash of file contents?

Reply:
We did remove the ID requirement, but I thought we settled on a random filename instead so people wouldn't mistakenly write code assuming the filename. Should I change it to be the content hash?

Reply:
I think hash is better than random as a default. The rest of the URI is unknown, so I think it is ok.

Reply:
ok, updated to use content hash